### DISCUSSION TITLE
Support for auto-generated SQL migrations in online mode

### DISCUSSION THREAD

**Discussion by basvandriel** [sqlalchemy/alembic#1545] [Ideas]
**Describe the use case**
Not everyone wants to use an ORM to add and edit database migrations. For enterprise graded production databases, reading raw SQL might even be more useful for database administrators. Alembic does currently not support that, other then the [offline mode](https://alembic.sqlalchemy.org/en/latest/offline.html). 

**Databases / Backends / Drivers targeted**
To start with, PostgreSQL

**Example Use**

After configuration, I foresee something like the following steps.
1. A new revision should be created `alembic revision --autogenerate --sql -m "Added account table"`, which will create one more SQL file containing the code for upgrading and downgrading the database. Similar to the regular migration files, this will also hold metadata.
2. Then, using the metadata, we can simply run `alembic upgrade head`, which will take the files into account.

**Comment by zzzeek** [comment]
you can put raw SQL in a migration file like this:


```py
def upgrade():
    op.get_bind().exec_driver_sql("""

CREATE TABLE ....

""")

```

however, the ask here seems to be that autogenerate would run normally, and create migration files, but instead of using alembic migration ops, would create a new .sql file with some migration that reads it.

We have had requests for migration files placed in .sql mode, for which it's easy enough to create migration files that consume the SQL files, but for autogenerate to create these files in the first place which are generated in any case from alembic's own operators, I dont see the point of that.

>  reading raw SQL might even be more useful for database administrators

absolutely, which is what ``--sql`` mode is for, run `alembic upgrade base:head --sql > myfile.sql` and hand off `myfile.sql` to your database administrators.   your DBAs are either reading your source repository directly, in which case it's pretty trivial to read what an alembic op does, or they're not, in which case you just give them the file.

**Reply by zzzeek** [reply]
it's likely that this use could also be produced using a [custom rewriter function](https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#customizing-revision-generation).  it would be a little involved since it would have to run the ops to generate the SQL into a separate file, but should be doable.   this can be rolled custom as a recipe and we can support adding such a recipe to our [docs](https://alembic.sqlalchemy.org/en/latest/cookbook.html)

**Reply by zzzeek** [reply]
here's that:

```py
import pathlib

from alembic.operations import Operations
from alembic.migration import MigrationContext


def write_sql_files(context, revision, directives):
    migrations = directives[0]

    for ops, op in [
        (migrations.upgrade_ops, "upgrades"),
        (migrations.downgrade_ops, "downgrades"),
    ]:
        version_path = context.environment_context.script.versions
        norm_path = pathlib.Path(version_path)
        with open(
            norm_path / f"{migrations.rev_id}_{op}.sql", "w"
        ) as output_file:
            opts = context.opts.copy()
            opts["as_sql"] = True
            opts["output_buffer"] = output_file
            sql_migration_context = MigrationContext.configure(
                context.connection, opts=opts
            )
            operations = Operations(sql_migration_context)

            stack = [ops]
            while stack:
                elem = stack.pop(0)

                if hasattr(elem, "ops"):
                    stack.extend(elem.ops)
                else:
                    operations.invoke(elem)

```

set this as the rewriter:

```py
def run_migrations_online() -> None:
    connectable = engine_from_config(
        config.get_section(config.config_ini_section, {}),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            process_revision_directives=write_sql_files,
        )

        with context.begin_transaction():
            context.run_migrations()

```

this traversal is based on the traverse we see at https://alembic.sqlalchemy.org/en/latest/cookbook.html#run-alembic-operation-objects-directly-as-in-from-autogenerate

**Reply by zzzeek** [reply]
the above just writes the files for readability purposes.  it could be extended to also rewrite the `MigrationScript` object in place to be code that reads the .sql files instead, or you could simply alter the `script.py.mako` script to hardcode to this routine.

**Reply by basvandriel** [reply]
While the above gives the user the expected output, I still think we should just use the `.sql` files instead of also generating the Python script. This so don't have any duplication and won't cause confusion which files are being used. Regarding the offline mode script generation, this could be handy when we want a full database SQL export of that specific revision. In this case, the export should be per revision. I'm open to work on this to also add it to the cookbook. Should we re-open the issue?

**Reply by zzzeek** [reply]
you can use the .sql files.  change your `script.py.mako` template to be hardcoded to reading the files, rather than using the ``${upgrades}`` or `${downgrades}` tokens.

there's no issue to be opened for adding doc recipes

**Reply by phiresky** [reply]
is there any way to get the actual script path within process_revision_directives? otherwise with `file_template` set the paths are awkward:
<img width="424" height="114" alt="image" src="https://github.com/user-attachments/assets/dc230138-84f5-4622-8eda-39862bf642d7" />

**Reply by zzzeek** [reply]
im not sure what you're asking, but all of that information is on each MigrationScript object: `.version_path`, `.rev_id`, `.message`.   `file_template` is on `context.script_directory.file_template` and you can also feed the three elements plus a date into `context.script_directory._rev_path()` to get the full path that will be generated.

**Reply by phiresky** [reply]
yes, I found `.version_path`, but it is None , at least at the stage the `process_revision_directives` is executed. `context` does not seem to have a `script_directory` attr, from looking at the code the `script_directory` is only passed into some of its methods. I did see `_rev_path` but I couldn't find access to either script_directory or any of its params apart from `rev_id`.

Thank you for your response!

**Reply by phiresky** [reply]
`_rev_path()` also takes a datetime parameter which in its normal use is generated directly without being stored, so even if I had the other parameters to the function I don't think it would be possible to use it to get the actual path that will be used since it has not been decided yet?

**Reply by zzzeek** [reply]
> context does not seem to have a script_directory attr, 

oh it passes a migrationcontext, not a revisioncontext... you should be able to find it in `context.opts['revision_context'].script_directory`.  this should likely not be buried like that

> _rev_path() also takes a datetime parameter which in its normal use is generated directly without being stored, 

so with that script_directory, you can also call `script_directory._geneate_create_date()` to get that

>  I don't think it would be possible to use it to get the actual path that will be used since it has not been decided yet?

in this convo we're working out a bit of code that we can turn into a utility function or an accessor on that context that's passed, like `context.get_revision_path(migration_script)` or something like that (maybe `migration_script.get_revision_path(context)` ?) , that would do this work and be available in a future release

**Reply by zzzeek** [reply]
> so with that script_directory, you can also call script_directory._geneate_create_date() to get that

with the caveat that it's called each time, so if your migration path had microseconds in it, or you called it often at 11:59:59 AM etc., that could be a problem.  there's not a workaround for that without changing the code.   The date to be used should be associted iwth the MigrationScript just once and re-used but it doesnt work like that at the moment.

**Reply by phiresky** [reply]
thank you! the non-deterministic timestamp is not great but it works! here's what I have now. I really just want the SQL files to be the one and only source of truth.

here's what I have now:

env.py:

```py
from typing import Any, cast

from alembic import context
from alembic.migration import MigrationContext
from alembic.operations import MigrateOperation, MigrationScript, Operations
from alembic.script import ScriptDirectory
# ...
# https://github.com/sqlalchemy/alembic/discussions/1545
# https://github.com/sqlalchemy/alembic/issues/609
def write_sql_files(
    context: MigrationContext,
    _revision: str | Iterable[str | None] | Iterable[str],
    directives: list[MigrationScript],
):
    assert len(directives) == 1
    migrations = directives[0]
    script_directory = cast(
        ScriptDirectory, context.opts["revision_context"].script_directory
    )
    # non-deterministic but best we can do
    create_date = script_directory._generate_create_date()  # noqa: SLF001
    # version_path = script_directory._singular_version_location
    assert context.environment_context is not None
    version_path = pathlib.Path(context.environment_context.script.versions)
    assert migrations.rev_id is not None
    rev_path = script_directory._rev_path(  # noqa: SLF001
        version_path, migrations.rev_id, migrations.message, create_date
    )

    for ops, op in [
        (migrations.upgrade_ops, "upgrades"),
        (migrations.downgrade_ops, "downgrades"),
    ]:
        sql_fname = rev_path.with_name(rev_path.name.removesuffix(".py") + f"_{op}.sql")
        with sql_fname.open("w") as output_file:
            opts = context.opts.copy()
            opts["as_sql"] = True
            opts["output_buffer"] = output_file
            sql_migration_context = MigrationContext.configure(
                context.connection, opts=opts
            )
            operations = Operations(sql_migration_context)
            assert ops is not None
            stack: list[MigrateOperation] = [ops]
            while stack:
                elem = stack.pop(0)
                assert elem is not None
                if hasattr(elem, "ops"):
                    stack.extend(cast(Any, elem).ops)
                else:
                    operations.invoke(elem)


def run_migrations_online() -> None:
    configuration = context.config.get_section(context.config.config_ini_section, {})
    if "sqlalchemy.url" not in configuration:
        configuration["sqlalchemy.url"] = os.environ["DATABASE_URL"]
    connectable = engine_from_config(
        configuration,
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            process_revision_directives=write_sql_files,
        )

        with context.begin_transaction():
            context.run_migrations()
```

and `script.py.mako`:

```py
"""${message}

Revision ID: ${up_revision}
Revises: ${down_revision | comma,n}
Create Date: ${create_date}

"""
from alembic import op
from pathlib import Path
import sqlalchemy as sa
${imports if imports else ""}

# revision identifiers, used by Alembic.
revision = ${repr(up_revision)}
down_revision = ${repr(down_revision)}
branch_labels = ${repr(branch_labels)}
depends_on = ${repr(depends_on)}

def file_path(op: str):
    file = Path(__file__)
    return file.with_name(file.name.removesuffix(".py") + f"_{op}.sql")

def upgrade():
    query = sa.text(file_path("upgrades").read_text())
    op.execute(query)

def downgrade():
    query = sa.text(file_path("downgrades").read_text())
    op.execute(query)
```

of course it would be great if doing this was simpler. I'm still not sure why we need python code at all for migrations, the (non-python) migration systems I know all work purely with SQL files, usually using the filename as the migration ID
